core/arch/arm/kernel/thread.c

```diff
@@ -253,6 +257,10 @@ void thread_set_exceptions(uint32_t exceptions)
 {
 	uint32_t daif = read_daif();
 
+	/* IRQ must not be unmasked while holding a spinlock */
+	if (!(exceptions & THREAD_EXCP_IRQ))
+		assert_have_no_spinlock();
```

Comment:
maybe the `have_no_spinlock()` and `spinlock_count_add()` could assert IRQ are masked from the CPSR/DAIF, before using core position?
obviously we can't call `thread_get/set_exceptions()` from these.